### ISSUE TITLE
Native symbolic integration layer (NSIL) for open-weight/local LLM agent architectures

### ISSUE THREAD

**Issue by rohanvinaik** [rohanvinaik/LintGate#129]
## Context

Issue #44 proposes cross-model MCP compatibility (adapters, schema validation, installer contracts) for broad interoperability across agent CLIs. That work treats LintGate as an external tool surface — model calls LintGate via MCP, gets results back, decides what to do.

For **open-weight / locally-hosted** model stacks, we can go further. Instead of treating LintGate as an external tool boundary, we can make its symbolic systems part of the agent's **native control architecture** — injecting state into context during inference, constraining generation against policy grammars, intercepting proposed actions before commit, and ultimately training models that internalize LintGate's symbolic discipline as first-order behavior.

This is a materially different design space from MCP. The existing codebase already has the pieces:

- **Symbolic state** is well-structured: `FunctionMutationState`, `SurvivorSite`, `CoverageDepth`, `SignalQuality`, `ConfidenceLevel` in `lintgate/mutation/state.py`; gate contracts in `gate_contract.yaml`; session memory in `lintgate/controlplane/session_memory.py`.
- **Host adapter protocol** exists: `lintgate/renderers/host_adapter.py` defines `HostCapabilities` with presets for Claude, Cursor, Copilot, Windsurf, Cline, Aider, and a generic `MCP_ONLY_CAPABILITIES` fallback.
- **Model calibration** exists: `lintgate/controlplane/model_profiles.py` supports arbitrary `model_key` registration with `resolve_model_key()` handling provider prefix mapping.
- **Hook system** is extensible: `hook_posttooluse.py` fires after tool use with stdin JSON → stdout JSON protocol, already enqueues mutation tasks non-blocking.
- **ControlPlane channels** run in parallel via the mesh orchestrator (`controlplane/runtime.py`) and produce machine-readable findings with theory codas.

**What's missing is the integration layer that connects these symbolic systems to local inference runtimes.** The state exists but isn't inference-ready. The adapters exist but don't speak to local model servers. The calibration system profiles models but doesn't constrain them. The hook system observes but doesn't intercept.

This issue defines the concrete engineering work to build that layer, organized as an integration tier progression from lightweight state injection through architecture-native training.

## Relationship to Existing Issues

- **#44** (cross-model MCP compatibility) is the portability layer — making LintGate install and validate correctly across agent CLIs. NSIL assumes #44's adapter abstractions exist and extends them deeper for local runtimes.
- **#102** (mutation-guided refactoring engine) and **#114** (connective tissue) built the symbolic analysis chain. NSIL makes that chain available as native model context, not just MCP tool responses.
- **#43** (system mutation detection) established the guardrail pattern for detecting unscoped side effects. NSIL's action interception layer (Gap D) generalizes this to constrained generation.

### Gap A: Symbolic state not inference-ready

The state structures (`FunctionMutationState`, gate contract snapshots, parity/drift state, risk classifications, unresolved blockers) are Python objects with MCP JSON serialization. They are designed for tool-call responses — verbose, nested, with metadata fields irrelevant to inference.

For injection into model context during generation, we need a **compact typed schema** that:
- Fits within a token budget (target: 200-500 tokens for full project state snapshot)
- Uses flat key-value pairs or short structured blocks, not deeply nested objects
- Includes only decision-relevant fields (survival rates, gate status, blocking findings — not run IDs or timestamps)
- Has a deterministic serialization that models can learn to condition on
- Supports incremental updates (delta encoding) for streaming contexts

### Gap B: No local runtime adapter abstraction

`HostCapabilities` in `host_adapter.py` models IDE/CLI feature detection (rules, hooks, MCP, frontmatter). It doesn't model **inference-loop integration points**: token streaming hooks, action proposal interception, state injection during generation, or policy-constrained decoding.

Local runtimes (Ollama, vLLM, llama.cpp, TGI, HuggingFace Inference Endpoints) each expose different APIs for these capabilities. There is no shared adapter interface that abstracts over them.

### Gap C: No policy grammar compiler

Gate contracts (`gate_contract.yaml`) and behavioral constraints (from `constraint_proposer.py`) exist as YAML/Python structures. For Tier 2 integration (constrained decoding), these need to compile into **grammar constraints** that local runtimes can enforce during token generation — e.g., GBNF grammars for llama.cpp, regex constraints for Outlines/vLLM, or logit processors for HuggingFace.

### Gap D: No action proposal interception

The current architecture is **observe-then-report**: code is generated, hooks fire, findings are surfaced. For Tier 3 integration (neuro-symbolic co-processor), we need **propose-then-verify**: the model proposes an action, the symbolic engine validates it against contracts and state, and the action is committed only if it passes.

The hook system (`hook_pretooluse.py` exists at 3.1KB) has the entry point but no verification loop — it's a constraint check, not a propose/verify/repair cycle.

### Gap E: No training data pipeline

ControlPlane traces, ship reports, behavioral session data, prediction accuracy logs, and constraint outcomes exist as JSON state files. None of this is formatted as training examples for fine-tuning open-weight models. There is no:
- Extraction pipeline from session artifacts → (prompt, completion, reward) triples
- Annotation schema for contract-adherent vs contract-violating actions
- Curriculum structure (basic gate compliance → specification-aware optimization → full symbolic reasoning)

### Gap F: No evaluation harness for native vs external integration

There is no way to measure whether deeper integration tiers actually improve outcomes. We need controlled benchmarks comparing:
- External MCP baseline (Tier 0) vs state-injected (Tier 1) vs grammar-constrained (Tier 2) vs co-processor (Tier 3)
- Metrics: policy compliance rate, latency per action, task completion rate, token cost, false-positive repair rate

---

## Implementation Plan

Ordered by integration tier — each phase builds on the previous and unlocks the next tier.

### Phase 1: Inference-Ready Symbolic State Schema (Gap A)

**Goal:** A compact, typed, token-budget-aware serialization of LintGate project state suitable for injection into model context during inference.

**Deliverables:**
1. Define `InferenceStateSnapshot` in new module `lintgate/nsil/state_schema.py`:
   ```python
   @dataclass
   class InferenceStateSnapshot:
       gate_status: str              # "green" | "red" | "yellow"
       blocking_findings: list[str]  # max 5, short descriptions
       mutation_summary: dict        # {fn: survival_rate} for top-5 hotspots
       active_constraints: list[str] # from constraint_proposer
       prediction_accuracy: float    # rolling accuracy from behavioral compass
       risk_level: str               # "low" | "medium" | "high"
       token_count: int              # self-measured serialized size
   ```
2. Add `serialize_compact(format: str)` supporting `"structured_text"` (for context injection), `"json_flat"` (for tool responses), and `"kv_pairs"` (for minimal-token formats).
3. Add projection functions that extract `InferenceStateSnapshot` from existing state sources:
   - `project_from_controlplane(run_result)` → snapshot
   - `project_from_session_memory(session)` → snapshot
   - `project_from_mutation_state(state_manager)` → snapshot
4. Add token budget enforcement: if serialized size exceeds budget, progressively drop lower-priority fields (findings → constraints → mutation detail).
5. Expose via MCP tool `nsil_state_snapshot(path, format, token_budget)`.

**Files:**
- New: `lintgate/nsil/__init__.py`, `lintgate/nsil/state_schema.py`
- Modified: `mcp_tools/` (new tool registration)
- Tests: `tests/test_nsil_state_schema.py`

**Acceptance:**
- Snapshot for a real project serializes to under 500 tokens in `structured_text` format.
- Progressive field dropping works: 200-token budget still produces valid, parseable output.
- Round-trip: `deserialize(serialize(snapshot)) == snapshot`.

---

### Phase 2: Local Runtime Adapter Interface (Gap B)

**Goal:** A shared abstraction for hooking into local inference loops, with one reference implementation.

**Deliverables:**
1. Define `LocalRuntimeAdapter` protocol in `lintgate/nsil/runtime_adapter.py`:
   ```python
   class LocalRuntimeAdapter(Protocol):
       def inject_state(self, snapshot: InferenceStateSnapshot) -> None: ...
       def register_action_hook(self, callback: ActionVerifier) -> None: ...
       def get_generation_stream(self) -> AsyncIterator[TokenEvent]: ...
       def apply_grammar_constraint(self, grammar: PolicyGrammar) -> None: ...
       def get_capabilities(self) -> RuntimeCapabilities: ...
   ```
2. Define `RuntimeCapabilities` (extends `HostCapabilities` with inference-specific flags):
   - `supports_state_injection: bool`
   - `supports_streaming_hooks: bool`
   - `supports_grammar_constraints: bool`
   - `supports_logit_processors: bool`
   - `max_context_tokens: int`
   - `api_protocol: str`  (openai-compat, native, custom)
3. Implement `OllamaAdapter` as reference (Ollama has the lowest barrier to entry, OpenAI-compatible API, widespread local use):
   - State injection via system prompt prepend
   - Generation streaming via `/api/generate` stream
   - No grammar constraint support initially (capability reports false)
4. Implement `VLLMAdapter` as second reference (vLLM supports guided generation natively):
   - State injection via system prompt
   - Grammar constraints via Outlines integration
   - Logit processor support via `SamplingParams`
5. Add adapter discovery: `detect_local_runtimes()` scans for running services on standard ports.

**Files:**
- New: `lintgate/nsil/runtime_adapter.py`, `lintgate/nsil/adapters/ollama.py`, `lintgate/nsil/adapters/vllm.py`
- Tests: `tests/test_nsil_adapters.py` (mock-based, no live runtime required)

**Acceptance:**
- `OllamaAdapter` successfully injects state snapshot into generation context via mock.
- `VLLMAdapter` reports grammar constraint capability as True.
- Adapter discovery correctly identifies running Ollama instance on default port.
- Adapters degrade gracefully when runtime is unavailable (clear error, not crash).

---

### Phase 3: Policy Grammar Compiler (Gap C)

**Goal:** Compile gate contracts and behavioral constraints into grammar rules that local runtimes can enforce during token generation.

**Deliverables:**
1. Define `PolicyGrammar` in `lintgate/nsil/grammar_compiler.py`:
   ```python
   @dataclass
   class PolicyGrammar:
       source_constraints: list[str]   # original constraint descriptions
       gbnf_rules: str | None          # for llama.cpp
       regex_pattern: str | None        # for Outlines/vLLM
       logit_bias: dict[int, float] | None  # for token-level bias
   ```
2. Implement constraint→grammar compilation for three constraint types:
   - **Action type constraints**: "do not run `rm -rf`" → grammar that rejects dangerous shell patterns in tool-call generation
   - **Gate status constraints**: "all tests must pass before push" → grammar enforcing verification steps before commit actions
   - **Scope constraints**: "only modify files in `src/`" → grammar restricting file path arguments
3. Compile from `gate_contract.yaml` required_checks and from `constraint_proposer.py` active constraints.
4. Support GBNF output (llama.cpp), regex output (Outlines/vLLM), and a human-readable explanation format.
5. Wire into `VLLMAdapter.apply_grammar_constraint()`.

**Files:**
- New: `lintgate/nsil/grammar_compiler.py`
- Modified: `lintgate/nsil/adapters/vllm.py`
- Tests: `tests/test_nsil_grammar.py`

**Acceptance:**
- Gate contract with 4 required checks compiles to a valid GBNF grammar.
- Regex output accepted by Outlines `RegexGuide` constructor (validated with import, not live run).
- Grammar correctly rejects a test case containing a prohibited action pattern.
- Compilation is deterministic: same inputs → same outputs.

---

### Phase 4: Action Proposal Interception Loop (Gap D)

**Goal:** A propose→verify→repair cycle where the symbolic engine validates model-proposed actions before they execute.

**Deliverables:**
1. Define `ActionProposal` and `VerificationResult` in `lintgate/nsil/action_verifier.py`:
   ```python
   @dataclass
   class ActionProposal:
       action_type: str          # "bash", "write", "edit", "tool_call"
       target: str               # file path or command
       content: str              # proposed content
       context: dict             # model's reasoning context

   @dataclass
   class VerificationResult:
       approved: bool
       violations: list[str]     # constraint violations found
       repairs: list[str]        # suggested repairs
       confidence: float
   ```
2. Implement verification pipeline that checks proposals against:
   - Active gate contract constraints
   - Behavioral constraints from `constraint_proposer`
   - Hygiene preconditions from `hygiene_check` logic
   - Scope boundaries (file paths, command classes)
3. Implement repair suggestion engine: when a proposal violates constraints, suggest minimal modifications that would pass.
4. Wire into `LocalRuntimeAdapter.register_action_hook()` so adapters can intercept generation and route through verification.
5. Add `nsil_verify_action(path, action_type, target, content)` MCP tool for manual/testing use.

**Files:**
- New: `lintgate/nsil/action_verifier.py`
- Modified: `lintgate/nsil/runtime_adapter.py`, `mcp_tools/`
- Tests: `tests/test_nsil_action_verifier.py`

**Acceptance:**
- Proposal containing `rm -rf /` is rejected with clear violation message.
- Proposal to write a file outside project scope is rejected.
- Repair suggestions for a rejected proposal are syntactically valid alternatives.
- Verification latency under 50ms for typical proposals (no LLM calls in the loop).

---

### Phase 5: Training Data Pipeline (Gap E)

**Goal:** Extract structured training examples from LintGate session artifacts for fine-tuning open-weight models.

**Deliverables:**
1. Define training example schema in `lintgate/nsil/training_data.py`:
   ```python
   @dataclass
   class TrainingExample:
       prompt: str                    # state snapshot + task description
       completion: str                # action taken
       reward: float                  # outcome signal (-1.0 to 1.0)
       labels: dict[str, str]         # contract_adherent, gate_outcome, etc.
       source: str                    # session_id, trace_id
   ```
2. Implement extraction pipelines from existing session artifacts:
   - `extract_from_controlplane_traces()` — ControlPlane run results → (state, action, outcome) triples
   - `extract_from_prediction_log()` — predictions + actual outcomes → calibration examples
   - `extract_from_constraint_outcomes()` — constraint checks + results → compliance examples
   - `extract_from_ship_reports()` — ship gate pass/fail → end-to-end task examples
3. Implement reward computation:
   - **Contract adherence**: +1.0 for gate-passing actions, -1.0 for violations, scaled by severity
   - **Cost-to-green**: reward inversely proportional to actions needed to reach passing gates
   - **Prediction accuracy**: +1.0 for correct predictions, -0.5 for incorrect
4. Add `nsil_extract_training_data(path, format, since)` MCP tool outputting JSONL.
5. Add curriculum ordering: basic compliance examples first, then optimization, then multi-step reasoning.

**Files:**
- New: `lintgate/nsil/training_data.py`
- Modified: `mcp_tools/`
- Tests: `tests/test_nsil_training_data.py`

**Acceptance:**
- Extraction from a real session produces at least 10 well-formed training examples.
- Reward signals are bounded and consistent (same outcome → same reward).
- JSONL output is compatible with standard fine-tuning frameworks (HuggingFace `datasets.load_dataset`).
- Curriculum ordering places compliance examples before optimization examples.

---

### Phase 6: Evaluation Harness (Gap F)

**Goal:** Controlled benchmarks comparing integration tiers on policy compliance, latency, task success, and token cost.

**Deliverables:**
1. Define evaluation protocol in `lintgate/nsil/eval_harness.py`:
   - Standard task set: 10 coding tasks with known gate requirements (test writing, bug fixing, refactoring, dependency updates)
   - Metric collectors: compliance rate, latency per action, total tokens, task completion, false-positive repairs
2. Implement tier runners:
   - `run_tier0(task)` — baseline MCP tool calls only
   - `run_tier1(task)` — state injection into context
   - `run_tier2(task)` — grammar-constrained generation
   - `run_tier3(task)` — full propose/verify/repair loop
3. Add comparison reporter producing markdown tables and delta analysis.
4. Add `nsil_benchmark(path, tiers, tasks, model)` MCP tool.

**Files:**
- New: `lintgate/nsil/eval_harness.py`, `lintgate/nsil/eval_tasks/` (task definitions)
- Modified: `mcp_tools/`
- Tests: `tests/test_nsil_eval_harness.py`

**Acceptance:**
- Evaluation runs to completion on at least 3 tasks across Tier 0 and Tier 1 with a local model.
- Metrics are collected consistently and comparison table is generated.
- No evaluation task requires internet access or paid API calls.

---

## Integration Tier Summary

| Tier | Name | Integration Point | Latency | Guarantee Level |
|------|------|-------------------|---------|-----------------|
| 0 | External MCP | Tool call/response | High (round-trip) | Advisory |
| 1 | State-Injected | System prompt prepend | Low (one-shot) | Contextual |
| 2 | Grammar-Constrained | Decoding constraints | Medium (per-token) | Structural |
| 3 | Neuro-Symbolic Co-Processor | Propose/verify loop | Medium (per-action) | Contractual |
| 4 | Architecture-Native Training | Loss/reward objectives | None (internalized) | Behavioral |

Phases 1-4 build Tiers 1-3. Phase 5 produces the data for Tier 4. Phase 6 measures the delta between tiers.

## Priority and Phasing

- **Phase 1** (state schema) is the foundation — everything else depends on compact, typed state being available. Highest payoff per effort.
- **Phase 2** (runtime adapters) is the integration surface — without it, nothing reaches a local model. Second priority.
- **Phase 3** (grammar compiler) is the first real differentiation from MCP — constrained decoding provides guarantees that tool calls cannot.
- **Phase 4** (action verifier) is the Tier 3 unlock — the propose/verify loop is where native integration provides a qualitative leap over external tools.
- **Phase 5** (training data) is the long-term investment — fine-tuned models that internalize LintGate discipline are the endgame.
- **Phase 6** (eval harness) should start early and run continuously — measuring each phase's impact guides prioritization.

Suggested execution order: 1 → 2 → 6 (baseline measurement) → 3 → 4 → 6 (delta measurement) → 5.

## Definition of Done

This issue is done when:
1. `InferenceStateSnapshot` serializes LintGate project state to under 500 tokens with progressive field dropping.
2. At least two local runtime adapters (Ollama + vLLM) implement the `LocalRuntimeAdapter` protocol with state injection.
3. Gate contracts compile to valid grammar constraints (GBNF + regex formats).
4. The propose/verify/repair loop rejects contract-violating actions with suggested repairs in under 50ms.
5. Training data extraction produces well-formed JSONL from real session artifacts.
6. Evaluation harness demonstrates measurable compliance improvement from Tier 0 → Tier 1 on at least 3 benchmark tasks.
7. All new functionality lives under `lintgate/nsil/` with deterministic test coverage and no regression to existing MCP tool contracts.

## Supersedes

This issue supersedes and closes **#45** (Exploratory: Native symbolic integration for open-weight/local LLM agent architectures), transforming its conceptual scope into a concrete implementation plan with defined gaps, phased deliverables, and measurable acceptance criteria.

**Comment by rohanvinaik** [comment]
Closed via audit: NSIL integration layer fully implemented. `action_verifier.py` (shared verification), `grammar_compiler.py` (PolicyGrammar with GBNF/regex), `streaming.py` (StreamingGuard with sliding window buffer), `state_schema.py` (InferenceStateSnapshot with 3 serialization formats), `runtime_adapter.py` (LocalRuntimeAdapter protocol + runtime discovery), `benchmark.py` (deterministic harness). Subsumed by super-issue #177.
